### PR TITLE
Add integration between autoclassification and manual starring.

### PR DESCRIPTION
**Pull request by jgraham**
This adds a new FailureClassification for autoclassified
intermittent. When a job is completely classified by the
autoclassifer, and it has the same number of structured and
unstructured error lines, it is marked as an autoclassified
intermittent.

Conversely, when there is exactly one structured and one unstructured
error line, the autoclassifier did not match the job, but has a
detector that could match the job, and the job is marked as
intermittent by a human, add a new autoclassification target
corresponding to the error line.

[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1051)

### CODE REVIEW

**Comment by jgraham** [comment]
Probably not ready to merge yet since it only passes some tests, I didn't actually try it. But this is a convenient way to share progress.

**Comment by maurodoglio** [comment]
Reviewed 2 of 9 files at r1, 10 of 10 files at r2.
Review status: 12 of 15 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

<sup>**[treeherder/model/derived/jobs.py, line 399 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fqUWEfIkuH6XS6GBS)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L399)):</sup>
This can we rewritten to: 
```python
intermittent_ids = FailureClassification.objects.filter(
    name__in=["intermittent", "intermittent needs filing"]).values_list('id', flat=True)
```


---

<sup>**[treeherder/model/derived/jobs.py, line 421 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fsFMiYbeAes9FLFTq)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L421)):</sup>
`s` missing in clasification


---

<sup>**[treeherder/model/derived/jobs.py, line 436 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fsfa_2qrY5kC7ywnk)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L436)):</sup>
The line above seems to be useless


---

<sup>**[treeherder/model/derived/jobs.py, line 440 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fuDa9h1ZlmCcrQPoO)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L440)):</sup>
do we want to have multiple classification objects associated to the same bug? If not, create_new_classification should renamed to something like set_classification and the body of that method should attempt to retrieve an existing classification related to that bug number.


---

<sup>**[treeherder/model/derived/jobs.py, line 450 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fvKFt9VBqymJPO6KG)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L450)):</sup>
I would use .get here and rename failure_lines to failure_line since you are looking for a single object.  You can except FailureLine.DoesNotExist and FailureLine.MultipleObjectsReturned to return early.


---

<sup>**[treeherder/model/derived/jobs.py, line 461 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fwxEQ_QfYy1uBzOg_)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L461)):</sup>
it would probably be more explicit here to use
```python
if any(detector(failure_lines) for detector in Matcher.objects.registered_detectors())
```
I'll leave it to your judgement


---

<sup>**[treeherder/model/derived/jobs.py, line 487 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0fy6I1lDX6P5oKPi4T)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L487)):</sup>
it looks like you only need a count here, not the entire list of failure lines.


---

<sup>**[treeherder/model/derived/jobs.py, line 523 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0g-s0ex9gcmxLV3VfN)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L523)):</sup>
why do you retrieve 2 objects here?


---

<sup>**[treeherder/model/models.py, line 589 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0g0XshhukWVB-LO1_E)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/models.py#L589)):</sup>
I'm not sure we want to create new classified failures for the same bug number here. Anyway, you can use either objects.create or objects.get_or_create to avoid the call to save.  


---

<sup>**[treeherder/model/models.py, line 592 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0g1F3crm5qtuxrapKo)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/models.py#L592)):</sup>
I would wrap this update and the next create in a transaction, so that you don't risk to leave the FailureLine object without a best match. 


---

<sup>**[ui/js/providers.js, line 147 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0g2IsBusRfNgYgpUgs)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/ui/js/providers.js#L147)):</sup>
can you please add a comment telling what classificationId 7 means?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

File: treeherder/model/models.py [hunk 1]
```diff
@@ -611,39 +621,68 @@ class Meta:
...
 
-    def _register(self, cls, dest):
-        if cls.__name__ in dest:
+    @classmethod
+    def _register(cls, matcher_cls, dest):
+        if matcher_cls.__name__ in dest:
             return dest[cls.__name__]
 
-        obj = Matcher.objects.get_or_create(name=cls.__name__)[0]
+        obj = Matcher.objects.get_or_create(name=matcher_cls.__name__)[0]
```
Comment (maurodoglio) [inline treeherder/model/models.py]: watch out, the value returned is a tuple

**Comment by maurodoglio** [comment]
Reviewed 1 of 9 files at r1, 6 of 10 files at r3, 2 of 2 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[tests/conftest.py, line 456 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E8wr2UFGeyIRnulMp)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/tests/conftest.py#L456)):</sup>
there's no need to save the repository, it was created with Repository.objects.create() in the test_repository fixture


---

<sup>**[treeherder/model/derived/jobs.py, line 523 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0g-s0ex9gcmxLV3VfN)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L523)):</sup>
I still don't get it. Why there's a limit of 35 elements when you only need one? There should never be more than one element here. All the bug suggestions should be in the same artifact.


---

<sup>**[treeherder/model/models.py, line 611 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E2Phkw09W2uzA4dhP)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/treeherder/model/models.py#L611)):</sup>
what about renaming this to 'get_best_for_line' ?


---

<sup>**[treeherder/model/models.py, line 615 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E2ES_D3VIFGKIY2IJ)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/treeherder/model/models.py#L615)):</sup>
You can pass the same arguments directly to .get(). Don't need to use filter here.


---

<sup>**[treeherder/model/models.py, line 725 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E6JIeEIJ2oBn_Ev9-)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/treeherder/model/models.py#L725)):</sup>
this change will probably require a new migration. can you check that using the following command?
```bash
python manage.py makemigrations
```
If it does generate a new migration file, you will need to add it to this patch


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by jgraham** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[tests/conftest.py, line 456 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E8wr2UFGeyIRnulMp)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/tests/conftest.py#L456)):</sup>
Well that's what I thought. And yet removing it makes tests fail. So perhaps something more significant is wrong.


---

<sup>**[treeherder/model/derived/jobs.py, line 523 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K0g-s0ex9gcmxLV3VfN)** ([raw file](https://github.com/mozilla/treeherder/blob/6b08baddd1e0a2045f1ba9555ed5c963aca8a1ed/treeherder/model/derived/jobs.py#L523)):</sup>
Oh, I see.


---

<sup>**[treeherder/model/models.py, line 725 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E6JIeEIJ2oBn_Ev9-)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/treeherder/model/models.py#L725)):</sup>
OK. I guess I'm surprised that something that doesn't actually affect the db requires a migration.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by maurodoglio** [comment]
Review status: 14 of 17 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

<sup>**[treeherder/model/models.py, line 725 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E6JIeEIJ2oBn_Ev9-)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/treeherder/model/models.py#L725)):</sup>
Yeah that's a bit surprising but I'm sure there's a good reason behind it.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by maurodoglio** [comment]
Review status: 14 of 17 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

<sup>**[tests/conftest.py, line 456 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E8wr2UFGeyIRnulMp)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/tests/conftest.py#L456)):</sup>
what the failure look like?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by jgraham** [comment]
Review status: 14 of 17 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

<sup>**[tests/conftest.py, line 456 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E8wr2UFGeyIRnulMp)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/tests/conftest.py#L456)):</sup>
The first one is:

_________________________________________________________________ test_classify_test_failure __________________________________________________________________

activate_responses = None, jm = <treeherder.model.derived.jobs.JobsModel object at 0xb61d954c>, test_project = 'test_treeherder'
test_repository = <Repository: test_treeherder development>, eleven_jobs_stored = None, initial_data = None
failure_lines = [<FailureLine: FailureLine object>, <FailureLine: FailureLine object>]
classified_failures = [<ClassifiedFailure: ClassifiedFailure object>, <ClassifiedFailure: ClassifiedFailure object>]

    def test_classify_test_failure(activate_responses, jm, test_project, test_repository,
                                   eleven_jobs_stored, initial_data, failure_lines,
                                   classified_failures):
        job = jm.get_job(2)[0]
    
        test_failure_lines = create_failure_lines(test_repository,
                                                  job["job_guid"],
                                                  [(test_line, {}),
                                                   (test_line, {"subtest": "subtest2"}),
                                                   (test_line, {"status": "TIMEOUT"}),
                                                   (test_line, {"expected": "ERROR"}),
                                                   (test_line, {"message": "message2"})])
    
        autoclassify(jm.project, job, test_failure_lines)
    
        expected_classified = test_failure_lines[:2]
        expected_unclassified = test_failure_lines[2:]
    
        for actual, expected in zip(expected_classified, classified_failures):
>           assert [item.id for item in actual.classified_failures.all()] == [expected.id]
E           assert [] == [1L]
E             Right contains more items, first extra item: 1L
E             Use -v to get the full diff

The second one is another case where the expected data looks wrong.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by maurodoglio** [comment]
Reviewed 1 of 3 files at r6.
Review status: 14 of 17 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[tests/conftest.py, line 456 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1051#-K1E8wr2UFGeyIRnulMp)** ([raw file](https://github.com/mozilla/treeherder/blob/96d5881419fe5e8b17abf9f9d81973bd88c236f6/tests/conftest.py#L456)):</sup>
I don't know what's wrong here. I looks like the tip of an iceberg.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by maurodoglio** [comment]
Reviewed 2 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1051)

**Comment by edmorley** [comment]
Please can the bug number be included in the commit message - makes git log easier :-)

**Pull request merged**